### DISCUSSION TITLE
How to select Bundle(s) from subquery?

### DISCUSSION THREAD

**Discussion by bmakan** [sqlalchemy/sqlalchemy#11202] [Usage Questions] [answered]
I'm struggling to select Bundle(s) from a subquery. It always adds new tables to the list of `FROM` clause instead of picking the columns from the subquery. I tried several techniques already without success: aliased (doesn't work with Bundles), select_from, with_entities...

E.g.:
```python
bundle = Bundle('mybundle', A.x, A.y)
q = session.query(bundle).filter(A.x > 5)
subq = q.subquery()

# Here I tried various approaches
q = session.query(bundle).select_from(subq)  # Tables joined in subq appeared in FROM clause
q = session.query(aliased(bundle, subq)).select_from(subq)  # Error: bundle cannot be used in aliased

q = session.query(*subq.c).select_from(subq)
dicts = [row.mybundle for row in q.all()]  # AttributeError: mybundle
```

**Comment by zzzeek** [comment, marked as answer]
hi, we can only work from examples of what you are trying to do.  

here's select bundle from a subquery, try running this


```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy.orm import Bundle
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    x: Mapped[int]
    y: Mapped[int]


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(x=1, y=2))
s.commit()


subq = select(A).where(A.x > 5).subquery()

stmt = select(Bundle("mybundle", subq.c.x, subq.c.y))

print(s.execute(stmt).all())
```

**Reply by bmakan** [reply]
Oh, I see. I have my own Bundle subclass which processes rows into dicts (with some special formatting rules). I tried to extract bundles from the subquery or make the bundle extract columns from the subquery. It didn't occur to me to construct a new bundle by passing in the subquery columns.

I ended up querying `Model.__table__`(s) and processing those into dicts.

Editing original question with wrong approaches.

**Reply by Alc-Alc** [reply]
@zzzeek apologies for necroing this thread, I have a quick question (happy to create a new discussion if needed)

Consider the example above, but with a change, the query is `subq = select(A, func.now()).subquery()` (notice I added one more column). I want to bundle all these columns (not just x and y), is `select(Bundle("mybundle", *subq.c))` the correct way? I am not sure about unpacking `subq.c`, is that ok or do you have better alternatives?

To phrase it differently, I am not sure if I even want a subquery, I just want to bundle an entire model with one more additional column. eg: `select(Bundle("mybundle", A, func.now()))` but that gives me an error `AttributeError: 'AnnotatedTable' object has no attribute '_label'`

**Reply by zzzeek** [reply]
you can unpack `subq.c` like that, that's fine

yes passing a whole `A` like that wont work right now
